rpc/transport/ws/server.go

```diff
@@ -141,54 +142,43 @@ func (wsc *serverWebSocketTransport) request(w http.ResponseWriter, r *http.Requ
 	}
 }
 
-func (wsc *serverWebSocketTransport) listenForClose(ctx context.Context, c *websocket.Conn, closeChan chan<- error) {
-	_, _, err := c.Read(ctx)
-	closeChan <- err
-	wsc.wg.Done()
-}
-
+var upgrader = websocket.Upgrader{} // use default options
 func (wsc *serverWebSocketTransport) subscribe(w http.ResponseWriter, r *http.Request) {
 	// TODO: We currently allow requests from any origins. We should probably use a whitelist.
-	opts := &websocket.AcceptOptions{InsecureSkipVerify: true}
-	c, err := websocket.Accept(w, r, opts)
+	upgrader.CheckOrigin = func(r *http.Request) bool { return true }
+
+	c, err := upgrader.Upgrade(w, r, nil)
 	if err != nil {
 		panic(err)
 	}
-	defer c.Close(websocket.StatusInternalError, "server initiated websocket close")
+
+	defer c.Close()
 	notificationChan := make(chan []byte)
 	key := strconv.Itoa(int(rand.Uint64()))
 	wsc.notificationListeners.Store(key, notificationChan)
 	defer wsc.notificationListeners.Delete(key)
 
-	// A client closes a connection by sending a message over the websocket
-	// This code converts the socket `Read` call to a channel.
-	// Ideally, all signals would be managed in the select statement below. But there is no channel API for the websocket read.
 	closeChan := make(chan error)
-	wsc.wg.Add(1)
-	go wsc.listenForClose(r.Context(), c, closeChan)
+
+	closeHandler := c.CloseHandler()
+	c.SetCloseHandler(func(code int, text string) error {
+		closeChan <- nil
```

Comment:
Does it matter that we're signalling on the `closeChan` before `closeHandler` completes?

Reply:
I can't think of any issues with that. Are you thinking about anything specific?

Reply:
I think that means we'll signal the `closeChan` (and return from `subscribe`) before the `upgrader` has finished executing `closeHandler()`? I'm not quite sure what the `c.CloseHandler` does so it might be ok, but in general I think we should only signal something is done after all execution is complete. This [ADR](https://github.com/statechannels/go-nitro/blob/7b8a110a9a0a302a254a2412510bcdc05fd9c482/.adr/0016-closing-go-routines.md#L83) might be relevant.